_posts/2022-09-06-Umesh-GSoC-on-ConscryptMode.md

```diff
@@ -0,0 +1,24 @@
+To be more consistent with Android, Robolectric needed to be updated to use Conscrypt as the default security provider.
+ 
+I was already well familiar with the technologies used in the project such as Java, Kotlin, Android Studio, Git and GitHub. I was interested in Android Developement and I was looking for an android project that uses Java as the primary language. I found Robolectric to be a perfect project to contribute to based on my skills, and I was ready to learn about testing concepts since Robolectric is a popular unit testing framework for Android. 
+
+At first, we thought we could switch Robolectric from BouncyCastle to Conscrypt directly. We quickly found we can't completely switch Robolectric to Conscrypt because there were some internal tests that needed BouncyCastle as a fallback security provider. After we added it in, we decided to create a ConscryptMode annotation that lets users choose to use Conscrypt or not. Robolectric can now use either Conscrypt and BouncyCastle or just BouncyCastle as the security provider. Robolectric will search for a requested security feature from Conscrypt first. If it does not support it, Robolectric will try BouncyCastle second.
+
+- If ConscryptMode is `ON`, it will install Conscrypt and BouncyCastle.
+- If ConscryptMode is `OFF`, it will only install BouncyCastle.
+
+I have learnt about unit testing like how to test whether a piece of code is working according to the expected behavior or not, writing tests for specific scenario, and providing self-explanatory names to methods so that other developers can understand it easily in the future.
+
+I learnt how to debug a large codebase like Robolectric and how to find the main cause behind an error/test case failure. I learnt about annotations like defining the annotation and how it works when we use it somewhere in the project, and how the compiler records these annotations to provide that information at times whenever needed. 
+
+When we contribute to an open-source project, we always need to check how our changes can affect other end users/projects that depend on it. Sometimes there may be cases where we can't fix/resolve issues because our project also depends on some other library/project. That might mean we are left to wait for a new update and hope that it fixes those errors. Like in our case, we used Conscrypt but it doesn't have support for Mac M1. We were left to either create our own custom Conscrypt or wait for an update from the official Conscrypt. It taught me that we should be aware of the impact that a project can have in a positive and negative way so that we can find the best solution that doesn't affect end users in a bad way.
```

Comment:
I have also learned about the tradeoffs that sometimes arise in software projects. For example, Conscrypt does not currently support Mac M1, and to some extent this is out of our control. We were aware of this issue yet decided to move forward with making Conscrypt the default security provider on Linux, Windows, and Mac x86_64. The increased fidelity for these environments outweighs the lack of M1 support. We also have opportunities to collaborate with Conscrypt for Mac M1 support, or explore alternatives like repackaging Conscrypt for Robolectric until Conscrypt supports M1.